### DISCUSSION TITLE
Why is "require-await" a controversial rule?

### DISCUSSION THREAD

**Discussion by ConnorSinnott** [eslint/eslint#18271] [General]
I often have nits with PRs where the "async" keyword is attached to function that is simply returning a promise without any "awaits." I was looking at implementing the "require-await" rule in our organization, but was surprised to see no "--fix" option. That led me to [this PR](https://github.com/eslint/eslint/pull/11167) which shows the rule is a bit controversial. 

Specifically @ljharb says:
> (note that personally I find the require-await to be a hugely dangerous rule that drastically misunderstands the nature of async functions in the language, and i don't recommend anyone enable it, and I'd prefer eslint never have included it in the first place)

Could we elaborate on this a bit? When I recommend developers remove the "async" keyword from a function returning a promise without awaiting, its because I know that all the async keyword is doing (correct me if I'm wrong) is adding a redundant step to the call stack.

**Comment by ljharb** [comment]
The `async` keyword serves a purpose on its own - it ensures the function can *never* do anything but return a promise. Specifically, it can never throw a sync exception. This prevents bugs and provides a valuable syntactic guarantee.

Unlike a generator that never yields, an async function that never awaits is still incredibly useful and valuable, and making the function "normal" is highly unlikely to be a good idea.

**Reply by nujhong** [reply]
I am surprised this didn't get more upvotes - the rule is very misleading imo.

See how both functions have the same signatures in typescript.

```ts
const THROW = true

function makePromise() {
  return new Promise<void>((resolve) => resolve());
}

// function fn(): Promise<void>
function fn() {
    if (THROW) {
        throw 'error'
    }
    
    return makePromise();
}

// function asyncFn(): Promise<void>
async function asyncFn() {
    if (THROW) {
        throw 'error'
    }

    return makePromise();
}

(function main() {
    asyncFn().catch((e) => {
        console.log("Caught:", e);
    });

    try {
        fn().catch((e) => {
            console.log("Never caught:", e);
        });
    } catch (e) {
        console.log("Uncaught:", e); //  👈 This is why it's dangerous
    }
})();
```

Playground [typescript](https://insiders.vscode.dev/tsplay/#code/MYewdgzgLgBAKgCQEoHkDqMC8MoCcCuApgFDEBm+YwUAluDALYCGA1oQAq4gM0SEAUAShgBvYjBi5CUfLjAwwhAO4xO3XoQA8ANxA0AJgD5+-KRBAAbbYWGZDkwuasDBg4gF9SFKrXpkwQqLiEjA0ZDD8iKhowmIh8TgAFlwqAOSEuFy4qcESnvG5DjJyjKwcXDx8QgDcHqRMEACeVDDe1HTyDc3AAGIBsYVhEVHoAwkhUMkgaRlZOfGehVLF8sxsapUutYv8bb6rTDT9QfFdVH1CAHTATFDAiSY2WPZx4zCgkJaElxYgAOb8ABEAGEmPg-okoAAuQEAGhgNlqC0EtUKeEaJze-iuNzuD34TzsmLeIQ+Tm+vwBgIAcoRrLh3mCIdC4QiUYVkUiQu5GXiIk9XuMyV8fv8gQBVKhMyEw+GIwqedyCGpAA) [eslint](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW5kZWY6IFwib2ZmXCIgKi9cblxuY29uc3QgVEhST1cgPSB0cnVlXG5cbmZ1bmN0aW9uIG1ha2VQcm9taXNlKCkge1xuICByZXR1cm4gbmV3IFByb21pc2U8dm9pZD4oKHJlc29sdmUpID0+IHJlc29sdmUoKSk7XG59XG5cbmZ1bmN0aW9uIGZuKCkge1xuICAgIGlmIChUSFJPVykge1xuICAgICAgICB0aHJvdyAnZXJyb3InXG4gICAgfVxuICAgIFxuICAgIHJldHVybiBtYWtlUHJvbWlzZSgpO1xufVxuXG5hc3luYyBmdW5jdGlvbiBhc3luY0ZuKCkge1xuICAgIGlmIChUSFJPVykge1xuICAgICAgICB0aHJvdyAnZXJyb3InXG4gICAgfVxuXG4gICAgcmV0dXJuIG1ha2VQcm9taXNlKCk7XG59XG5cbihmdW5jdGlvbiBtYWluKCkge1xuICAgIGFzeW5jRm4oKS5jYXRjaCgoZSkgPT4ge1xuICAgICAgICBjb25zb2xlLmxvZyhcIkNhdWdodDpcIiwgZSk7XG4gICAgfSk7XG5cbiAgICB0cnkge1xuICAgICAgICBmbigpLmNhdGNoKChlKSA9PiB7XG4gICAgICAgICAgICBjb25zb2xlLmxvZyhcIk5ldmVyIGNhdWdodDpcIiwgZSk7XG4gICAgICAgIH0pO1xuICAgIH0gY2F0Y2ggKGUpIHtcbiAgICAgICAgY29uc29sZS5sb2coXCJVbmNhdWdodDpcIiwgZSk7XG4gICAgfVxufSkoKTsiLCJvcHRpb25zIjp7InJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjpbImVycm9yIl0sImZvci1kaXJlY3Rpb24iOlsiZXJyb3IiXSwiZ2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjpbImVycm9yIl0sIm5vLWNhc2UtZGVjbGFyYXRpb25zIjpbImVycm9yIl0sIm5vLWNsYXNzLWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb21wYXJlLW5lZy16ZXJvIjpbImVycm9yIl0sIm5vLWNvbmQtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb25zdGFudC1iaW5hcnktZXhwcmVzc2lvbiI6WyJlcnJvciJdLCJuby1jb25zdGFudC1jb25kaXRpb24iOlsiZXJyb3IiXSwibm8tY29udHJvbC1yZWdleCI6WyJlcnJvciJdLCJuby1kZWJ1Z2dlciI6WyJlcnJvciJdLCJuby1kZWxldGUtdmFyIjpbImVycm9yIl0sIm5vLWR1cGUtYXJncyI6WyJlcnJvciJdLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOlsiZXJyb3IiXSwibm8tZHVwZS1lbHNlLWlmIjpbImVycm9yIl0sIm5vLWR1cGUta2V5cyI6WyJlcnJvciJdLCJuby1kdXBsaWNhdGUtY2FzZSI6WyJlcnJvciJdLCJuby1lbXB0eSI6WyJlcnJvciJdLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tZW1wdHktcGF0dGVybiI6WyJlcnJvciJdLCJuby1lbXB0eS1zdGF0aWMtYmxvY2siOlsiZXJyb3IiXSwibm8tZXgtYXNzaWduIjpbImVycm9yIl0sIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6WyJlcnJvciJdLCJuby1mYWxsdGhyb3VnaCI6WyJlcnJvciJdLCJuby1mdW5jLWFzc2lnbiI6WyJlcnJvciJdLCJuby1nbG9iYWwtYXNzaWduIjpbImVycm9yIl0sIm5vLWltcG9ydC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW52YWxpZC1yZWdleHAiOlsiZXJyb3IiXSwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOlsiZXJyb3IiXSwibm8tbG9zcy1vZi1wcmVjaXNpb24iOlsiZXJyb3IiXSwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tbmV3LW5hdGl2ZS1ub25jb25zdHJ1Y3RvciI6WyJlcnJvciJdLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6WyJlcnJvciJdLCJuby1vYmotY2FsbHMiOlsiZXJyb3IiXSwibm8tb2N0YWwiOlsiZXJyb3IiXSwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjpbImVycm9yIl0sIm5vLXJlZGVjbGFyZSI6WyJlcnJvciJdLCJuby1yZWdleC1zcGFjZXMiOlsiZXJyb3IiXSwibm8tc2VsZi1hc3NpZ24iOlsiZXJyb3IiXSwibm8tc2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6WyJlcnJvciJdLCJuby1zcGFyc2UtYXJyYXlzIjpbImVycm9yIl0sIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjpbImVycm9yIl0sIm5vLXVuZGVmIjpbImVycm9yIl0sIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjpbImVycm9yIl0sIm5vLXVucmVhY2hhYmxlIjpbImVycm9yIl0sIm5vLXVuc2FmZS1maW5hbGx5IjpbImVycm9yIl0sIm5vLXVuc2FmZS1uZWdhdGlvbiI6WyJlcnJvciJdLCJuby11bnNhZmUtb3B0aW9uYWwtY2hhaW5pbmciOlsiZXJyb3IiXSwibm8tdW51c2VkLWxhYmVscyI6WyJlcnJvciJdLCJuby11bnVzZWQtcHJpdmF0ZS1jbGFzcy1tZW1iZXJzIjpbImVycm9yIl0sIm5vLXVudXNlZC12YXJzIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtYmFja3JlZmVyZW5jZSI6WyJlcnJvciJdLCJuby11c2VsZXNzLWNhdGNoIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtZXNjYXBlIjpbImVycm9yIl0sIm5vLXdpdGgiOlsiZXJyb3IiXSwicmVxdWlyZS15aWVsZCI6WyJlcnJvciJdLCJ1c2UtaXNuYW4iOlsiZXJyb3IiXSwidmFsaWQtdHlwZW9mIjpbImVycm9yIl0sInJlcXVpcmUtYXdhaXQiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9LCJzb3VyY2VUeXBlIjoibW9kdWxlIn0sInBhcnNlciI6IkB0eXBlc2NyaXB0LWVzbGludC9wYXJzZXIifX19)
